### PR TITLE
Feature: Added EVM transfer/fetch limit

### PR DESCRIPTION
**Pull request by Janislav**
# Pull Request

Closes: PRO-1619

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have written sufficient tests.
- [x] I have written and tested required migrations.
- [x] I have updated documentation where appropriate.

## Summary

*Please include a succinct description of the purpose and content of the PR. What problem does it solve, and how? Link issues, discussions, other PRs, and anything else that will help the reviewer.*

#### Non-Breaking changes

If this PR includes non-breaking changes, select the `non-breaking` label. On merge, CI will automatically cherry-pick the commit to a PR against the release branch.

### CODE REVIEW

**Comment by Janislav** [comment]
After a discussion with @albert-llimos we came up with the added values for ETH which are:

- transfer_limit -> 50
- fetches_limit -> 20
- ccm_limit -> None

Unit/integration tests as well as bouncer is working. Any other ideas how we can check for any bad surprises @dandanlen ? I assume another level of verification is when it is out on a test net.

**Review by dandanlen — COMMENTED** [review]
The main thing to test would be to try to add more transfers/fetches than the limit and check that it actually works (ie we don't hit any edge cases). I guess this covered already in the unit tests?

**Comment by Janislav** [comment]
I don't think it is tested at all at the moment. Also not for other chains. I think it makes sense then to write unit tests in the ingress/egress pallet against a mock implementation just to test the basic behaviour.

**Comment by albert-llimos** [comment]
LGTM, I'll let Dan to the final review and approve.

**Review by albert-llimos — APPROVED** [review]

File: state-chain/runtime/src/lib.rs [hunk 1]
```diff
@@ -365,7 +367,7 @@ impl pallet_cf_ingress_egress::Config<Instance1> for Runtime {
 	type FeePayment = Flip;
 	type SwapRequestHandler = Swapping;
 	type AssetWithholding = AssetBalances;
-	type FetchesTransfersLimitProvider = NoLimit;
+	type FetchesTransfersLimitProvider = EthereumLimit;
```
Comment (kylezs) [inline state-chain/runtime/src/lib.rs:370]: Should the same limit be applied to Arbitrum @albert-llimos ?

Comment (albert-llimos) [reply state-chain/runtime/src/lib.rs:370]: yes definitely

Comment (kylezs) [reply state-chain/runtime/src/lib.rs:370]: would rename to EvmLimit then, and apply to Arbitrum too @Janislav

Comment (Janislav) [reply state-chain/runtime/src/lib.rs:370]: Really? Last time we talked about it, we said it will be most likely different limits for Arbitrum 🤔

File: state-chain/pallets/cf-ingress-egress/src/tests.rs [hunk 2]
```diff
@@ -1671,3 +1671,140 @@ fn safe_mode_prevents_deposit_channel_creation() {
...
+
+		assert_eq!(
+			schedule_egresses.len(),
+			transfer_limits + 1,
+			"Wrong amount of scheduled egresses!"
+		);
+
+		IngressEgress::on_finalize(1);
+
+		let schedule_egresses = ScheduledEgressFetchOrTransfer::<Test, ()>::get();
```
Comment (dandanlen) [inline state-chain/pallets/cf-ingress-egress/src/tests.rs:1697]: nit: this variable should be `sheduled_egresses`.

File: state-chain/pallets/cf-ingress-egress/src/tests.rs [hunk 3]
```diff
@@ -1671,3 +1671,140 @@ fn safe_mode_prevents_deposit_channel_creation() {
...
+
+		assert_eq!(schedule_egresses.len(), EXCESS_TRANSFERS, "Wrong amount of left egresses!");
+
+		IngressEgress::on_finalize(2);
+
+		let schedule_egresses = ScheduledEgressFetchOrTransfer::<Test, ()>::get();
+
+		assert_eq!(schedule_egresses.len(), 0, "Left egresses have been fully processed!");
+
+		MockFetchesTransfersLimitProvider::enable_limits(false);
```
Comment (dandanlen) [inline state-chain/pallets/cf-ingress-egress/src/tests.rs:1707]: Disabling it at the end is not really necessary.

File: state-chain/pallets/cf-ingress-egress/src/mock_eth.rs [hunk 4]
```diff
@@ -110,6 +112,44 @@ impl NetworkEnvironmentProvider for MockNetworkEnvironmentProvider {
 
 impl_mock_runtime_safe_mode! { ingress_egress_ethereum: PalletSafeMode<()> }
 
+thread_local! {
+	pub static USE_LIMITS: RefCell<bool> = RefCell::new(false);
+}
+
+pub struct MockFetchesTransfersLimitProvider;
```
Comment (dandanlen) [inline state-chain/pallets/cf-ingress-egress/src/mock_eth.rs:119]: Why define this here instead of in the trait crate, same as all the other mocks?

File: state-chain/pallets/cf-ingress-egress/src/tests.rs [hunk 5]
```diff
@@ -1671,3 +1671,140 @@ fn safe_mode_prevents_deposit_channel_creation() {
...
+
+		let schedule_egresses = ScheduledEgressFetchOrTransfer::<Test, ()>::get();
+
+		assert_eq!(schedule_egresses.len(), EXCESS_TRANSFERS, "Wrong amount of left egresses!");
+
+		IngressEgress::on_finalize(2);
+
+		let schedule_egresses = ScheduledEgressFetchOrTransfer::<Test, ()>::get();
+
+		assert_eq!(schedule_egresses.len(), 0, "Left egresses have been fully processed!");
```
Comment (dandanlen) [inline state-chain/pallets/cf-ingress-egress/src/tests.rs:1705]: nit: the assert message should give additional information for when the condition fails, not when it succeeds.

File: state-chain/pallets/cf-ingress-egress/src/tests.rs [hunk 6]
```diff
@@ -1671,3 +1671,140 @@ fn safe_mode_prevents_deposit_channel_creation() {
...
+
+		let schedule_egresses = ScheduledEgressFetchOrTransfer::<Test, ()>::get();
+
+		assert_eq!(schedule_egresses.len(), EXCESS_FETCHES, "Wrong amount of left egresses!");
+
+		IngressEgress::on_finalize(2);
+
+		let schedule_egresses = ScheduledEgressFetchOrTransfer::<Test, ()>::get();
+
+		assert_eq!(schedule_egresses.len(), 0, "Left egresses have been fully processed!");
```
Comment (dandanlen) [inline state-chain/pallets/cf-ingress-egress/src/tests.rs:1754]: Same as above. 
Was this meant to be `"Left egresses have not been fully processed!"`?

File: state-chain/runtime/src/chainflip.rs [hunk 7]
```diff
@@ -941,3 +941,19 @@ impl FetchesTransfersLimitProvider for SolanaLimit {
...
+
+pub struct EvmLimit;
+impl FetchesTransfersLimitProvider for EvmLimit {
+	fn maybe_transfers_limit() -> Option<usize> {
+		Some(50)
+	}
+
+	fn maybe_ccm_limit() -> Option<usize> {
+		// For ccm calls we don't batch
+		None
```
Comment (dandanlen) [inline state-chain/runtime/src/chainflip.rs:953]: We don't batch but it might be worth limiting this anyway since each ccm requires a signature ceremony... will think about it.

**Review by dandanlen — COMMENTED** [review]
Thanks, test cases are fine, just some cosmetic issues.

File: state-chain/pallets/cf-ingress-egress/src/mock_eth.rs (hunk 4 above)
Comment (Janislav) [reply state-chain/pallets/cf-ingress-egress/src/mock_eth.rs:119]: Because it's only used in this pallet at the moment and also only used by this pallet.

File: state-chain/pallets/cf-ingress-egress/src/tests.rs (hunk 3 above)
Comment (Janislav) [reply state-chain/pallets/cf-ingress-egress/src/tests.rs:1707]: It could be since if not the next test that gets executed would run against the enabled limits which might be wrong (depends on the test). If things getting shifted around in the future, this could lead to confusion.

File: state-chain/pallets/cf-ingress-egress/src/tests.rs [hunk 8]
```diff
@@ -1671,3 +1671,140 @@ fn safe_mode_prevents_deposit_channel_creation() {
...
+		MockFetchesTransfersLimitProvider::enable_limits(true);
+
+		const EXCESS_TRANSFERS: usize = 1;
+		let transfer_limits = MockFetchesTransfersLimitProvider::maybe_transfers_limit().unwrap();
+
+		for _ in 1..=transfer_limits + EXCESS_TRANSFERS {
+			assert_ok!(IngressEgress::schedule_egress(ETH_ETH, 1_000, ALICE_ETH_ADDRESS, None));
+		}
+
+		let sheduled_egresses = ScheduledEgressFetchOrTransfer::<Test, ()>::get();
```
Comment (kylezs) [inline state-chain/pallets/cf-ingress-egress/src/tests.rs:1687]:
```suggestion
		let scheduled_egresses = ScheduledEgressFetchOrTransfer::<Test, ()>::get();
```
think @dandanlen made a typo in his comment...

File: state-chain/pallets/cf-ingress-egress/src/tests.rs (hunk 3 above)
Comment (dandanlen) [reply state-chain/pallets/cf-ingress-egress/src/tests.rs:1707]: That's not really how it works. Every test (everything inside the `#[test] fn x() {..}` runs isolated from each other with fresh state.

File: state-chain/pallets/cf-ingress-egress/src/mock_eth.rs (hunk 4 above)
Comment (dandanlen) [reply state-chain/pallets/cf-ingress-egress/src/mock_eth.rs:119]: Ok but the convention we (try to) follow is to define the mocks where the trait is defined. This works well and I don't see a good reason to break with that convention here.

File: state-chain/pallets/cf-ingress-egress/src/tests.rs (hunk 8 above)
Comment (dandanlen) [reply state-chain/pallets/cf-ingress-egress/src/tests.rs:1687]: 🙈 yep sorry

File: state-chain/runtime/src/chainflip.rs (hunk 7 above)
Comment (dandanlen) [reply state-chain/runtime/src/chainflip.rs:953]: I've opened an issue to come back to this.

**Review by dandanlen — APPROVED** [review]

**Comment by kylezs** [comment]
Seems to have been a case of: https://linear.app/chainflip/issue/PRO-1627/fix-rpc-unavailable-upgrade-test-flakiness
